iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -1291,329 +1338,409 @@ static bool isSystemProperty(const char* tokenData)
-    // Codes_SRS_IOTHUB_TRANSPORT_MQTT_COMMON_31_062: [ If IoTHubTransport_MQTT_Common_DoWork receives a malformatted inputQueue, it shall fail ]
+    const char* result;
+    const char* inputNameStart;
+    const char* inputNameEnd;
+    char* inputNameCopy = NULL;
 
-    int result = MU_FAILURE;
-    int number_tokens_read = 0;
-
-    STRING_TOKENIZER_HANDLE token_handle = STRING_TOKENIZER_create_from_char(topic_name);
-    if (token_handle == NULL)
+    if (((inputNameStart = strchr(propertiesStart, TOPIC_SLASH)) == NULL) || (*(inputNameStart + 1) == '\0'))
     {
-        LogError("STRING_TOKENIZER_create_from_char failed\n");
-        result = MU_FAILURE;
+        LogError("Cannot find '/' to mark beginning of input name");
+        result = NULL;
     }
     else
     {
-        STRING_HANDLE token_value;
-        if ((token_value = STRING_new()) == NULL)
+        inputNameStart++;
+        if ((inputNameEnd = strchr(inputNameStart, TOPIC_SLASH)) == NULL)
         {
-            LogError("Failed allocating token_value");
+            LogError("Cannot find '/' after input name");
+            result = NULL;
         }
-        else
+        else 
         {
-            while (STRING_TOKENIZER_get_next_token(token_handle, token_value, "/") == 0)
+            size_t inputNameLength = inputNameEnd - inputNameStart;
+            if ((inputNameCopy = malloc(inputNameLength + 1)) == NULL)
             {
-                number_tokens_read++;
-                if (number_tokens_read == (slashes_to_reach_input_name + 1))
+                LogError("Cannot allocate input name");
+                result = NULL;
+            }
+            else
+            {
+                memcpy(inputNameCopy, inputNameStart, inputNameLength);
+                inputNameCopy[inputNameLength] = '\0';
+
+                if (IoTHubMessage_SetInputName(iotHubMessage, inputNameCopy) != IOTHUB_MESSAGE_OK)
                 {
-                    if ((IOTHUB_MESSAGE_OK != IoTHubMessage_SetInputName(IoTHubMessage, STRING_c_str(token_value))))
-                    {
-                        LogError("Failed adding input name to msg");
-                        result = MU_FAILURE;
-                    }
-                    else
-                    {
-                        result = 0;
-                    }
-                    break;
+                    LogError("Failed adding input name to msg");
+                    result = NULL;
+                }
+                else
+                {
+                    result = inputNameEnd + 1;
                 }
             }
         }
-        STRING_delete(token_value);
-
-        if (number_tokens_read != (slashes_to_reach_input_name + 1))
-        {
-            LogError("Not enough '/' to contain input name.  Got %d, need at least %d", number_tokens_read, (slashes_to_reach_input_name + 1));
-            result = MU_FAILURE;
-        }
-        STRING_TOKENIZER_destroy(token_handle);
     }
 
+    free(inputNameCopy);
```

Comment:
Nit: should move this free to inside the block where `inputNameCopy` becomes relevant. Otherwise it could be a moot call to free.

Reply:
i.e., to the block after malloc succeeds.